### DISCUSSION TITLE
Vídeo 07 - Desenvolvimento

### DISCUSSION THREAD

**Discussion by thpoiani** [thpoiani/lab-quarkus#9] [General]
Olá! Nesse vídeo montamos um sistema simples de mensageria com Redis PUBSUB.

Quais outros serviços de mensageria você conhece? RabbitMQ, SNS/SQS, Kafka, ... ?

E quais são os principais motivos de usar tal serviço em comparação com REST? (_hehe essa não é a pergunta ideal, que tal: sincronicidade vs assincronicidade?_)
Tenho uma referência muito interessante sobre isso e **recomendo fortemente que assista esse vídeo**: https://developertoarchitect.com/lessons/lesson137.html ⭐

**Comment by didifive** [comment]
No caso do Azure tem o Azure EventHub que é como se fosse o Kafka para o Azure Cloud, eu conheci também o ActiveMQ com os serviços de filas e tópicos.

Em relação ao REST com Mensageria, pensando em uma situação com aspecto assíncrono em que uma aplicação está esperando um recurso ter sido processado e atualizado por outra aplicação externa.

- No REST a aplicação que espera essa atualização teria que ativamente ficar realizando chamadas constantes para a outra aplicação até obter uma resposta, nestes casos acabam sendo necessários agendamentos e intervalos para chamadas e ainda com o planejamento para que as chamadas não sejam excessivas para evitar consumir recursos desnecessários ao mesmo tempo que o intervalo ou agendamento não ultrapasse tempo que possa atrasar desnecessariamente o processo da própria aplicação ou outras que dependem desta informação.
- Em Mensageria, a aplicação que espera o processamento e atualização pode se subscrever em uma fila que a outra aplicação irá publicar a resposta do processo, assim não é preciso criar agendamentos ou intervalos, podendo conhecer o resultado do processamento e atualização do recurso praticamente em tempo real.

**Comment by mauromortoni** [comment]
Usar serviços de mensageria em comparação com APIs REST pode oferecer diversas vantagens, dependendo das necessidades da aplicação e da arquitetura do sistema. Aqui estão os principais motivos para considerar serviços de mensageria:

1. Desacoplamento
- Independência: os serviços de mensageria permitem que os produtores e consumidores de mensagens operem de forma independente. Mudanças em um serviço não afetam diretamente o outro, facilitando a manutenção e a evolução da aplicação.

2. Escalabilidade
- Escalabilidade Horizontal: mensageria permite que você escale os serviços de forma independente, aumentando a capacidade de processamento sem impactar outros componentes.
- Balanceamento de Carga: vários consumidores podem processar mensagens de uma fila, ajudando a balancear a carga de trabalho.

3. Resiliência e Tolerância a Falhas
- Persistência de Mensagens: mensagens podem ser armazenadas em filas, garantindo que não sejam perdidas mesmo que o consumidor esteja temporariamente fora do ar.
- Retry e Dead Letter Queues: mensagens que falham podem ser reprocessadas ou enviadas para "filas de erro" para análise e tratamento posterior.

4. Processamento Assíncrono
- Desempenho Melhorado: com serviços de mensageria, os produtores podem continuar a operação sem esperar que o consumidor processe a mensagem, resultando em um fluxo de trabalho mais eficiente.
- Processamento em Lote: mensagens podem ser agrupadas e processadas em lotes, otimizando o uso de recursos.

5. Comunicação em Tempo Real
- Eventos em Tempo Real: sistemas de mensageria são ideais para cenários onde eventos precisam ser comunicados imediatamente a vários consumidores, como em sistemas de monitoramento ou aplicações de IoT.

6. Flexibilidade
- Vários Consumidores: várias aplicações podem consumir a mesma mensagem simultaneamente, permitindo que diferentes serviços realizem ações com base no mesmo evento.
- Pub/Sub: modelos de publicação/inscrição permitem que novos consumidores sejam adicionados sem alterar a lógica dos produtores.

7. Gerenciamento de Fluxo de Trabalho
- Orquestração de Serviços: sistemas de mensageria podem ser usados para orquestrar processos complexos, onde diferentes serviços precisam se comunicar e cooperar.

8. Melhor Gerenciamento de Carga
- Back Pressure: serviços de mensageria podem implementar back pressure, onde o sistema pode controlar a taxa de produção de mensagens em resposta à carga do consumidor.

9. Facilidade de Integração
- Interoperabilidade: serviços de mensageria podem facilitar a integração entre diferentes sistemas e tecnologias, tornando a comunicação mais flexível.

Considerações
Embora APIs REST sejam eficazes para muitas interações de serviço a serviço, os serviços de mensageria oferecem uma série de benefícios, especialmente em arquiteturas que requerem escalabilidade, resiliência e desacoplamento. A escolha entre REST e mensageria deve ser baseada nos requisitos específicos do sistema e nos padrões de comunicação desejados.
